crates/runner-nordic/src/storage.rs

```diff
@@ -90,20 +93,107 @@ impl Storage {
         unsafe { &*self.ptr }
     }
 
+    pub fn ptr(&self) -> *const u8 {
+        self.ptr as *const u8
+    }
+
     pub fn len(&self) -> usize {
         self.ptr.len()
     }
+}
 
-    pub const ERASE_SIZE: usize = PAGE_SIZE;
-    pub fn erase(&self, offset: usize, length: usize) -> StorageResult<()> {
-        let mut helper = Helper::new(self);
-        helper.nvmc().erase(offset as u32, (offset + length) as u32).map_err(convert)
+pub struct StorageWriter {
+    storage: Storage,
+    // None if not running.
+    dry_run: Option<bool>,
+    // offset + buffer.len() <= storage.len() && offset % WORD_SIZE == 0
+    offset: usize,
+    // buffer.len() < WORD_SIZE
+    buffer: Vec<u8>,
+}
+
+impl StorageWriter {
+    pub fn new(storage: Storage) -> Result<Self, (Storage, Error)> {
+        Ok(StorageWriter {
+            storage,
+            dry_run: None,
+            offset: 0,
+            buffer: Vec::with_capacity(WORD_SIZE),
+        })
     }
 
-    pub const WRITE_SIZE: usize = <Nvmc<NVMC>>::WRITE_SIZE;
-    pub fn write(&self, offset: usize, data: &[u8]) -> StorageResult<()> {
-        let mut helper = Helper::new(self);
-        helper.nvmc().write(offset as u32, data).map_err(convert)
+    pub fn dry_run(&self) -> Result<bool, Error> {
+        self.dry_run.ok_or(Error::user(Code::InvalidState))
+    }
+
+    pub fn start(&mut self, dry_run: bool) -> Result<(), Error> {
+        self.dry_run = Some(dry_run);
```

Comment:
We should check if `dry_run.is_some()` in which case we should reset the other fields.